contracts/components/staking/SlashingController.sol

```diff
@@ -0,0 +1,415 @@
+            .EMPTY_MACHINE
+            .addEdgeTransition(UNDEFINED, CREATED)
+            .addEdgeTransition(CREATED, DISMISSED)
+            .addEdgeTransition(CREATED, REJECTED)
+            .addEdgeTransition(CREATED, IN_REVIEW)
+            .addEdgeTransition(IN_REVIEW, REVIEWED)
+            .addEdgeTransition(IN_REVIEW, REVERTED)
+            .addEdgeTransition(REVIEWED, EXECUTED)
+            .addEdgeTransition(REVIEWED, REVERTED);
+    }
+
+    /**
+     * @notice Initializer method, access point to initialize inheritance tree.
+     * @param __manager address of AccessManager.
+     * @param __router address of Router.
+     */
+    function initialize(
+        address __manager,
+        address __router,
+        ISlashingExecutor __executor,
+        FortaStakingParameters __stakingParameters,
+        uint256 __depositAmount,
+        uint256 __slashPercentToProposer,
+        bytes32[] calldata __slashPenaltyIds,
+        SlashPenalty[] calldata __slashPenalties
+    ) public initializer {
+        __AccessManaged_init(__manager);
+        __Routed_init(__router);
+        __UUPSUpgradeable_init();
+
+        _setSlashingExecutor(__executor);
+        _setStakingParametersManager(__stakingParameters);
+        _setDepositAmount(__depositAmount);
+        _setSlashPercentToProposer(__slashPercentToProposer);
+        _setSlashPenalties(__slashPenaltyIds, __slashPenalties);
+    }
+
+    // Proposal LifeCycle
+    /**
+     * @notice Creates a slash proposal pointing to a slashable subject. To do so, the proposer must provide a FORT deposit and present evidence.
+     * @param _subjectType type of the subject.
+     * @param _subjectId ERC721 registry id of the stake subject.
+     * @param _penaltyId if of the SlashPenalty to inflict upon the subject if the proposal goes through.
+     * @param _evidence IPFS hashes of the evidence files, proof of the subject being slash worthy.
+     * @return proposalId the proposal identifier.
+     */
+    function proposeSlash(
+        uint8 _subjectType,
+        uint256 _subjectId,
+        bytes32 _penaltyId,
+        string[] calldata _evidence
+    ) external onlyValidSlashPenaltyId(_penaltyId) onlyValidSubjectType(_subjectType) returns (uint256 proposalId) {
+        if (!stakingParameters.isRegistered(_subjectType, _subjectId)) revert NonRegisteredSubject(_subjectType, _subjectId);
+        if (stakingParameters.totalStakeFor(_subjectType, _subjectId) == 0) revert ZeroAmount("subject stake");
+        Proposal memory slashProposal = Proposal(_subjectId, _msgSender(), _penaltyId, _subjectType);
+        SafeERC20.safeTransferFrom(depositToken, _msgSender(), address(this), depositAmount);
+        _proposalIds.increment();
+        proposalId = _proposalIds.current();
+        deposits[proposalId] = depositAmount;
+        proposals[proposalId] = slashProposal;
+        emit DepositSubmitted(proposalId, _msgSender(), depositAmount);
+        _transition(proposalId, CREATED);
```

Comment:
`_transition` contains a "check", and according to "checks-effects-interactions" we should call it at the beginning of the function. However, it also contains an "effect", so it's not that clear cut. If there is no reentrancy risk, the pattern doesn't make a difference, but it's usually better to stick to the pattern always, just in case.

Reply:
I moved `_transition` up in `markAsInReviewSlashProposal`, but in the case of `proposeSlash` I think there is no reentrancy issue:
- Incrementing `proposalId` first means you will be creating a new proposal, not modifying an existing one
- Each proposal creation requires a deposit, so it does not make sense to call it more than once if you don't mean it